src/basic/alloc-util.c

```diff
@@ -64,7 +65,7 @@ void* greedy_realloc(void **p, size_t *allocated, size_t need, size_t size) {
                 return NULL;
 
         *p = q;
-        *allocated = newalloc;
+        *allocated = _unlikely_(size == 0) ? newalloc : malloc_usable_size(q) / size;
```

Comment:
Do valgrind and addresssanitizer and such know that this is OK to access the bytes outside of the requested size? OK, it seems that both valgrind and asan cause malloc_usable_size() to always return the requested size. So I guess this can go in, although I'm not fully convinced this will not tickle the ire of some other checker.